### PR TITLE
Draft: Work/winvfs tests with helper

### PR DESCRIPTION
**Pull request by erikjv**

### CODE REVIEW

File: test/testutils/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,9 @@
+add_executable(test_helper test_helper.cpp)
+target_link_libraries(test_helper PUBLIC Qt5::Core libsync)
+
 add_library(syncenginetestutils STATIC syncenginetestutils.cpp testutils.cpp)
 target_link_libraries(syncenginetestutils PUBLIC owncloudCore Qt5::Test)
+target_compile_definitions(syncenginetestutils PRIVATE TEST_HELPER_EXE="$<TARGET_FILE:test_helper>")
```
Comment (TheOneRing) [inline test/testutils/CMakeLists.txt:6]: as everything ends up in build/bin we could relay on the path?

File: src/libsync/discovery.cpp [hunk 2]
```diff
@@ -719,7 +719,7 @@ void ProcessDirectoryJob::processFileAnalyzeLocalInfo(
 
     if (dbEntry.isValid()) {
         bool typeChange = localEntry.isDirectory != dbEntry.isDirectory();
-        if (!typeChange && localEntry.isVirtualFile) {
+        if (!typeChange && localEntry.isVirtualFile && dbEntry._modtime == localEntry.modtime && dbEntry._fileSize == localEntry.size) {
```
Comment (dragotin) [inline src/libsync/discovery.cpp:722]: since the "modtime and size haven't changed"-check is used twice I suggest to create an bool as `typeChange` above - might increase readability a bit.

**Review by dragotin — COMMENTED** [review]
just a small comment, no full review.

File: test/testsyncmove.cpp [hunk 3]
```diff
@@ -33,6 +32,11 @@ struct OperationCounter {
             return nullptr;
         };
     }
+
+    friend inline QDebug operator<<(QDebug dbg, const OperationCounter &oc)
```
Comment (TheOneRing) [inline test/testsyncmove.cpp:36]: off topic but I was recently wondering which version the preferable is.

`QDebug &operator<<(QDebug &dbg,`
 or 
`QDebug operator<<(QDebug &dbg,`

File: test/testsyncmove.cpp [hunk 4]
```diff
@@ -83,32 +87,55 @@ class TestSyncMove : public QObject
...
+    }
+
     void testRemoteChangeInMovedFolder()
     {
+        QFETCH_GLOBAL(Vfs::Mode, vfsMode);
+        QFETCH_GLOBAL(bool, filesAreDehydrated);
+
         // issue #5192
-        FakeFolder fakeFolder{ FileInfo{ QString(), { FileInfo{ QStringLiteral("folder"), { FileInfo{ QStringLiteral("folderA"), { { QStringLiteral("file.txt"), 400 } } }, QStringLiteral("folderB") } } } } };
+        FakeFolder fakeFolder(FileInfo { QString(), { FileInfo { QStringLiteral("folder"), { FileInfo { QStringLiteral("folderA"), { { QStringLiteral("file.txt"), 400 } } }, QStringLiteral("folderB") } } } }, vfsMode, filesAreDehydrated);
```
Comment (TheOneRing) [inline test/testsyncmove.cpp:113]: Why the change of { to (

File: test/testutils/syncenginetestutils.cpp [hunk 5]
```diff
@@ -101,13 +115,106 @@ void DiskFileModifier::mkdir(const QString &relativePath)
...
+        HelperProcess helper(_rootDir, _processArguments);
+        _processArguments.clear(); // the helper has a copy, we clean it now for the next run
+        while (!helper.finished) {
+            QThread::currentThread()->eventDispatcher()->processEvents(QEventLoop::AllEvents);
+            helper.waitForFinished();
+        }
+        if (!helper.succeeded) {
+            return false;
+        }
+        return ff.syncOnce();
```
Comment (TheOneRing) [inline test/testutils/syncenginetestutils.cpp:184]: Hm the implicit syncOnce will result in local==remote state, if nothing goes wrong. Is that really what we want here?

File: test/testutils/syncenginetestutils.cpp [hunk 6]
```diff
@@ -101,13 +115,106 @@ void DiskFileModifier::mkdir(const QString &relativePath)
...
+
+    do {
+        // process any pending events
+        QThread::currentThread()->eventDispatcher()->processEvents(QEventLoop::AllEvents);
+        // wait a bit to get the helper started with it's work
+        QThread::msleep(50);
+        // process any output from the helper
+        QThread::currentThread()->eventDispatcher()->processEvents(QEventLoop::AllEvents);
+        // now we might need to sync
+        ff.syncOnce();
```
Comment (TheOneRing) [inline test/testutils/syncenginetestutils.cpp:212]: I guess you need to explain the implicit syncs to me 😅 (again)

File: test/testutils/syncenginetestutils.h [hunk 7]
```diff
@@ -578,7 +585,14 @@ class FakeFolder
         return execUntilFinished();
     }
 
+    bool applyLocalModifications() Q_REQUIRED_RESULT
```
Comment (TheOneRing) [inline test/testutils/syncenginetestutils.h:588]: We use c++17 on master so I guess we should start using https://en.cppreference.com/w/cpp/language/attributes/nodiscard
